src/intrinsics.cpp

```diff
@@ -364,6 +367,22 @@ static Value *emit_unbox(jl_codectx_t &ctx, Type *to, const jl_cgval_t &x, jl_va
     }
     else {
         Instruction *load;
+        if (p->getType() != ptype && isa<AllocaInst>(p)) {
+            // LLVM's mem2reg can't handle coercion if the load/store type does
+            // not match the type of the alloca. As such, it is better to
+            // perform the load using the alloca's type and then perform the
+            // appropriate coercion manually.
+            AllocaInst *AI = cast<AllocaInst>(p);
+            Type *AllocType = AI->getAllocatedType();
+            if (!AI->isArrayAllocation() && !AllocType->isAggregateType() && !AllocType->isVectorTy()) {
```

Comment:
maybe test for what it can handle directly?

```
(AllocType->isFloatingPointTy() || AllocType->isIntegerTy() || AllocType->isPointerTy()) &&
(to->isFloatingPointTy() || AllocType->isIntegerTy() || AllocType->isPointerTy())
```